### PR TITLE
Detect cleanups/v4

### PR DESCRIPTION
**Pull request by victorjulien**
#9949 rebased

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 16884

**Review by jasonish — APPROVED** [review]
In general I think some commits could use a little more description.

> detect/dns.query: use new content inspect entry

Why?

> detect/content-inspect: localize recursion counting

Again, perhaps a little more detail.

Given there is no ticket for these.

File: src/tests/detect-http-uri.c [hunk 1]
```diff
@@ -6793,8 +6613,6 @@ static void DetectHttpUriRegisterTests (void)
 
     UtRegisterTest("UriTestSig28", UriTestSig28);
     UtRegisterTest("UriTestSig29", UriTestSig29);
-    UtRegisterTest("UriTestSig30", UriTestSig30);
-    UtRegisterTest("UriTestSig31", UriTestSig31);
```
Comment (jufajardini) [inline src/tests/detect-http-uri.c:6797]: Just to double check: is this the commit that added similar logic elsewhere? https://github.com/OISF/suricata/pull/9974/commits/32b4423d454e174ace618ac874d6f4ee920407ce

**Review by jufajardini — COMMENTED** [review]
Mostly looks ok to me, have one double check comment and a question about initialization of a struct value inline.

File: src/detect-engine-content-inspection.h [hunk 2]
```diff
@@ -28,7 +28,7 @@
 /** indication to content engine what type of data
  *  we're inspecting
  */
-enum {
+enum DetectContentInspectionType {
```
[the remark this answers is not in the document]
Comment (victorjulien) [reply src/detect-engine-content-inspection.h:31]: These days I'm finding things that are more explicit easier to work with. Typedef hides important detail.

File: src/detect-engine-content-inspection.c [hunk 3]
```diff
@@ -569,17 +569,16 @@ static int DetectEngineContentInspectionInternal(DetectEngineThreadCtx *det_ctx,
...
 
         /* if we have dce enabled we will have to use the endianness
          * specified by the dce header */
         if ((bmd->flags & DETECT_BYTEMATH_FLAG_ENDIAN) && endian == (int)EndianDCE &&
                 flags & (DETECT_CI_FLAGS_DCE_LE | DETECT_CI_FLAGS_DCE_BE)) {
-
             /* enable the endianness flag temporarily.  once we are done
              * processing we reset the flags to the original value*/
-            endian |= (uint8_t)((flags & DETECT_CI_FLAGS_DCE_LE) ? LittleEndian : BigEndian);
+            endian = (uint8_t)((flags & DETECT_CI_FLAGS_DCE_LE) ? LittleEndian : BigEndian);
```
[the remark this answers is not in the document]
Comment (victorjulien) [reply src/detect-engine-content-inspection.c:581]: it's a minor clarification. The resulting value is the same.

File: src/detect-engine-content-inspection.c [hunk 4]
```diff
@@ -95,18 +105,17 @@
...
         const enum DetectContentInspectionType inspection_mode)
 {
     SCEnter();
     KEYWORD_PROFILING_START;
 
-    det_ctx->inspection_recursion_counter++;
-
-    if (unlikely(det_ctx->inspection_recursion_counter == de_ctx->inspection_recursion_limit)) {
+    ctx->recursion.count++;
+    if (unlikely(ctx->recursion.count == ctx->recursion.limit)) {
```
[the remark this answers is not in the document]
Comment (victorjulien) [reply src/detect-engine-content-inspection.c]: both `DetectEngineContentInspectionBuffer` and `DetectEngineContentInspection` set it

[the remark this answers is not in the document]
Comment (victorjulien) [reply src/detect-engine-content-inspection.c]: the `ctx` is shared between consecutive recursive calls then, it's actually meant to count and limit that recursion

File: src/detect-engine-content-inspection.c [hunk 5]
```diff
@@ -714,6 +714,25 @@ bool DetectEngineContentInspection(DetectEngineCtx *de_ctx, DetectEngineThreadCt
...
 }
 
+/** \brief wrapper around DetectEngineContentInspectionInternal to return true/false only
+ *
+ *  \param smd sigmatches to evaluate
+ */
+bool DetectEngineContentInspectionBuffer(DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx,
+        const Signature *s, const SigMatchData *smd, Packet *p, Flow *f, const InspectionBuffer *b,
+        const enum DetectContentInspectionType inspection_mode)
+{
```
[the remark this answers is not in the document]
Comment (victorjulien) [reply src/detect-engine-content-inspection.c:751]: I've update the commit msg. It's a convenience function that keeps the internals of InspectionBuffer hidden from the callers.

File: src/tests/detect-http-uri.c (hunk 1 above)
Comment (victorjulien) [reply src/tests/detect-http-uri.c:6797]: heh, no, it's in a branch that is still WIP. I'll drop this commit from here.

**Comment by victorjulien** [comment]
replaced by #9986

**Pull request closed without merging**